nixos/modules/services/audio/botamusique.nix

```diff
@@ -0,0 +1,120 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.botamusique;
+
+  format = pkgs.formats.ini {};
+  configFile = format.generate "botamusique.ini" cfg.settings;
+in
+{
+  options.services.botamusique = {
+    enable = mkEnableOption "botamusique";
+
+    host = mkOption {
+      type = types.str;
+      default = "localhost";
+      example = "mumble.example.com";
+      description = "Hostname of the mumble server to connect to.";
+    };
+
+    port = mkOption {
+      type = types.port;
+      default = 64738;
+      description = "Port of the mumble server to connect to.";
+    };
+
+    username = mkOption {
+      type = types.str;
+      default = "botamusique";
+      description = "Name the bot should appear with.";
+    };
+
+    comment = mkOption {
+      type = types.str;
+      default = "Hi, I'm here to play radio, local music or youtube/soundcloud music. Have fun!";
+      description = "Comment displayed for the bot.";
+    };
+
+    channel = mkOption {
+      type = types.str;
+      description = "Channel to join on connect.";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.botamusique;
+      description = "The botamusique package to use.";
+    };
+
+    settings = mkOption {
+      type = format.type;
+      default = {};
+      description = ''
+        Your <filename>configuration.ini</filename> as a Nix attribute set. Look up
+        possible options in the <link xlink:href="https://github.com/azlux/botamusique/blob/master/configuration.example.ini">configuration.example.ini</link>.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.botamusique.settings = {
+      server = {
+        host = lib.mkDefault cfg.host;
+        port = lib.mkDefault cfg.port;
+        channel = lib.mkDefault cfg.channel;
+      };
+      bot = {
+        username = lib.mkDefault cfg.username;
+        comment = lib.mkDefault cfg.comment;
+      };
```

Comment:
You can use [freeform modules](https://github.com/NixOS/nixpkgs/pull/82743) for this, perfect use case for it.

Reply:
Does that mean I will loose the toplevel options? Or do I use their value in the default of the nested freeform option?

Reply:
You would move all the toplevel option definitions into the `settings` option. So it would be e.g. `services.botamusique.settings.bot.username`. Then you won't have to do the `mkDefault` thing here, and can define defaults with the options `default` argument.

Reply:
That sounds like it breaks discoverability big time.

Reply:
Is your last paragraph a question? If so, I'm not really sure what you're asking

Reply:
I've reworked this use the freeform type.